app/scripts/controllers/transactions/tx-state-manager.test.js

```diff
@@ -614,6 +631,136 @@ describe('TransactionStateManager', function () {
       );
       assert.equal(result[1].id, 2, 'early txs truncated');
     });
+
+    it('cuts off entire groups of transactions by nonce when adding new transaction', function () {
+      const limit = txStateManager.txHistoryLimit;
+      // In this test case the earliest two transactions are a dropped attempted ether send and a
+      // following cancel transaction with the same nonce. these two transactions should be dropped
+      // together as soon as the 11th unique nonce is attempted to be added, mocked here with limit + 1
+      const txs = generateTransactions(limit + 2, {
+        chainId: currentChainId,
+        to: VALID_ADDRESS,
+        from: VALID_ADDRESS_TWO,
+        nonce: (i) => (i === 1 ? `0` : `${i}`),
+        status: (i) =>
+          i === 0
+            ? TRANSACTION_STATUSES.DROPPED
+            : TRANSACTION_STATUSES.CONFIRMED,
+        type: (i) =>
+          i === 1 ? TRANSACTION_TYPES.CANCEL : TRANSACTION_STATUSES.SENT_ETHER,
+      });
+      txs.forEach((tx) => txStateManager.addTransaction(tx));
+      const result = txStateManager.getTransactions();
+      assert.equal(result.length, limit, `limit of ${limit} is enforced`);
+      assert.notEqual(result[0].id, 0, 'first tx should be removed');
+      assert.notEqual(
+        result[0].status,
+        TRANSACTION_STATUSES.DROPPED,
+        'first tx should not be dropped',
+      );
+      assert.notEqual(
+        result[1].type,
+        TRANSACTION_TYPES.CANCEL,
+        'Cancel tx should be dropped',
```

Comment:
It might be better to assert that the dropped and cancel transactions are _nowhere_ in the list of transactions. Just to ensure they're really gone.

If we had accidentally removed just the first transaction, this assertion would still pass because the cancel transaction isn't at index 1 anymore.

Reply:
done here and in the nonce+network test